### PR TITLE
Add post reading time and word count

### LINKED ISSUE #862: Add post reading time
#### Expected behavior

User can see an estimated time to read the post.

### PR DESCRIPTION
**Pull request by Kennybll**
Fixes #862.
Changes:
 - Adds Word Count and Reading Time estimate to the editor.
 - Adds NPM package "reading-time".
 - Adds Reading Time Estimate for Posts.

### CODE REVIEW

**Comment by Kennybll** [comment]
@bonustrack

**Comment by jm90m** [comment]
Hi thanks for the contribution,

Idk how useful this would be for the user if this is only in the editor.... Usually in blog websites, the time to read is on the article itself for users to know how long it will take to read the article..

See in medium:

screen shot 2017-11-21 at 8 45 06 am
screen shot 2017-11-21 at 8 45 09 am

**Comment by bonustrack** [comment]
I think word count make sense in the editor, and reading time in the post.

**Comment by Kennybll** [comment]
So add reading time to post, should I keep it on the editor as well?

**Comment by Kennybll** [comment]
Also, should I add reading time to the preview, full post, or both? And should there be a setting?

**Comment by bonustrack** [comment]
@Kennybll We don't need it on the preview and I don't think we need a setting for it.

**Comment by Kennybll** [comment]
Okay, should I commit to here or add another pull request?

**Comment by Kennybll** [comment]
@bonustrack @jm90m

**Comment by jm90m** [comment]
you can just commit here and make updates here. There's no point in making another PR. Maybe rename the PR to to whatever it is your changing & edit description.

**Comment by Kennybll** [comment]
Okay I added the Reading Time Estimate to the Posts Page.

**Comment by Kennybll** [comment]
I can make a copy/rewrite for busy.org under the helpers folder. @jm90m

**Comment by bonustrack** [comment]
@jm90m that was my recommendation to use this library. I've looked the options and this one seem to be the best to do the job. It's a really small job so i'm not surprise it's not been updated for a year and i don't see an issue about it. I prefer to use a library and not make our codebase more complex or bigger.

**Comment by jm90m** [comment]
@bonustrack sounds good, we can keep the library then.

**Comment by Kennybll** [comment]
Okay.

**Comment by Kennybll** [comment]
@bonustrack Hold on before you possibly merge. I looked at how many words reading-time says there is. I think it maybe adding tags or something, because on one post it had 3 links and 2 photos, yet when I debugged it said 7 words. That's incorrect. I think I can add the options parameter with a function than skips over the tags. I don't know. I'll look into it if you want.

**Comment by Kennybll** [comment]
Nevermind it handle markdown link fine.

File: src/components/Editor/Editor.js [hunk 1]
```diff
@@ -13,8 +15,14 @@ import EditorToolbar from './EditorToolbar';
 import Action from '../Button/Action';
 import Body, { remarkable } from '../Story/Body';
 import './Editor.less';
+import { getWordCount } from '../../reducers';
```
Comment (Sekhmet) [inline src/components/Editor/Editor.js]: We import styles last.

File: src/components/Editor/Editor.js [hunk 2]
```diff
@@ -13,8 +15,14 @@ import EditorToolbar from './EditorToolbar';
...
 
 @injectIntl
+@connect(
```
Comment (Sekhmet) [inline src/components/Editor/Editor.js]: I would rather keep Editor as a state-less component, so use connect from its parent and just pass it by props.

File: src/components/Editor/Editor.js [hunk 3]
```diff
@@ -34,6 +42,7 @@ class Editor extends React.Component {
     onError: PropTypes.func,
     onImageInserted: PropTypes.func,
     onImageInvalid: PropTypes.func,
+    wordCount: PropTypes.oneOf(['on', 'off']),
```
Comment (Sekhmet) [inline src/components/Editor/Editor.js]: Functions are last in prop-types.

File: src/components/Editor/Editor.js [hunk 4]
```diff
@@ -584,6 +594,11 @@ class Editor extends React.Component {
                 />
               )}
             </label>
+            {this.props.wordCount === 'on' &&
+              <label htmlFor="reading_time" className="Editor__readingTime">
+                <FormattedMessage id="reading_time" defaultMessage={`${readingTime(this.state.contentHtml).words} Words / ${readingTime(this.state.contentHtml).text}`} />
```
Comment (Sekhmet) [inline src/components/Editor/Editor.js]: `reading-time(...).text` is not translatable. We should make it translatable some way.

**Comment by Kennybll** [comment]
Sorry for all the commits, I'm try to get through the linting process.

**Comment by Sekhmet** [comment]
You can check for lint errors locally using `npm run lint` and `npm run test`.
There is precommit hook so they should be run before every commit.

**Comment by Kennybll** [comment]
@bonustrack @jm90m, @Sekhmet  thinks we shouldn't have a setting, what do you think?

**Comment by jm90m** [comment]
I don't think a settings option for this is necessary, usually users will always want to see their word count.

**Comment by Kennybll** [comment]
@bonustrack @jm90m @Sekhmet, that should do it, I think.

File: src/components/Editor/Editor.js [hunk 5]
```diff
@@ -584,6 +585,16 @@ class Editor extends React.Component {
                 />
               )}
             </label>
+            <label htmlFor="reading_time" className="Editor__readingTime">
+              <FormattedMessage
+                id="reading_time"
+                defaultMessage={'{words} Words / {min} min read'}
```
Comment (bonustrack) [inline src/components/Editor/Editor.js]: `Words` should be lowercased

File: src/components/Story/StoryFull.js [hunk 6]
```diff
@@ -266,6 +267,12 @@ class StoryFull extends React.Component {
       <div className="StoryFull">
         {replyUI}
         <h1 className="StoryFull__title">{post.title}</h1>
+        <h2 className="StoryFull__readting_time">
+          <FormattedMessage
+            id="reading_time"
+            defaultMessage={'{min} min read'}
+            values={{ min: Math.ceil(readingTime(post.body).minutes) }}
+          /></h2>
```
Comment (bonustrack) [inline src/components/Story/StoryFull.js]: `</h2>` should be closed in a new line

File: src/post/Write/Write.js [hunk 7]
```diff
@@ -19,6 +19,7 @@ import {
   getDraftPosts,
   getIsEditorLoading,
   getIsEditorSaving,
+  getWordCount,
```
Comment (bonustrack) [inline src/post/Write/Write.js]: We don't need reducer here, it seem that the method doesnt exist anymore.

File: src/post/Write/Write.js [hunk 8]
```diff
@@ -69,6 +72,7 @@ class Write extends React.Component {
     saveDraft: () => {},
     notify: () => {},
     replace: () => {},
+    wordCount: 'off',
```
Comment (bonustrack) [inline src/post/Write/Write.js]: If there is no "on" we dont need this line.

**Comment by Kennybll** [comment]
There @bonustrack

**Comment by bonustrack** [comment]
@Kennybll 
- Can we have this style for the time to read:
image
Using a bullet similar than `CommentFooter__bullet` class
- It would be nice to hide the reading time indicator if the content is less than 1000 characters.

**Comment by Sekhmet** [comment]
I don't think that removing indicator for short posts is useful. It might be confusing for users to see that some posts have it and some posts don't.

**Comment by Sekhmet** [comment]
@Kennybll can you run `npm run test` and `npm run lint` before committing? This would help you find problems locally, without need of pushing it to repo.

**Comment by Kennybll** [comment]
There we go @bonustrack @Sekhmet. The reason I wasn't using local lint was because my command prompt kept saying "Expected linebreaks to be 'LF' but found 'CRLF'" but I fixed that for the future.

**Comment by bonustrack** [comment]
@Kennybll Word count on editor should be aligned to right 
image

**Comment by Kennybll** [comment]
@bonustrack Woops when I changed Editor____readingTime to Editor__reading_time, I didn't change Editor.less, sorry It should work now.

File: src/components/Editor/Editor.js [hunk 9]
```diff
@@ -584,6 +585,16 @@ class Editor extends React.Component {
                 />
               )}
             </label>
+            <label htmlFor="reading_time" className="Editor__reading_time">
+              <FormattedMessage
```
Comment (Sekhmet) [inline src/components/Editor/Editor.js]: You have to add translations here as well:
https://github.com/busyorg/busy/blob/new-design/src/locales/en.json

**Comment by Kennybll** [comment]
@Sekhmet  added it to en.json

**Pull request merged**